core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttribute.java

```diff
@@ -276,61 +275,74 @@ public void setIfAbsent(Map<ExecutionAttribute<?>, Object> attributes, T value)
         }
     }
 
+    /**
+     * An implementation of {@link ValueStorage} that is backed by a different execution attribute in the provided
+     * attributes map (mirrors its value), and maps (updates) to another attribute.
+     */
     private static final class MappedValueStorage<T, U> implements ValueStorage<T> {
-        private final Supplier<ExecutionAttribute<U>> realAttributeSupplier;
-        private final AtomicReference<T> previousValue = new AtomicReference<>();
+        private final Supplier<ExecutionAttribute<T>> backingAttributeSupplier;
+        private final Supplier<ExecutionAttribute<U>> attributeSupplier;
         private final BiFunction<T, U, T> readMapping;
         private final BiFunction<U, T, U> writeMapping;
 
         private MappedValueStorage(MappedAttributeBuilder<T, U> builder) {
-            this.realAttributeSupplier = Validate.paramNotNull(builder.realAttributeSupplier, "realAttributeSupplier");
+            this.backingAttributeSupplier = Validate.paramNotNull(builder.backingAttributeSupplier, "backingAttributeSupplier");
+            this.attributeSupplier = Validate.paramNotNull(builder.attributeSupplier, "attributeSupplier");
             this.readMapping = Validate.paramNotNull(builder.readMapping, "readMapping");
             this.writeMapping = Validate.paramNotNull(builder.writeMapping, "writeMapping");
         }
 
         @SuppressWarnings("unchecked") // Safe because of the implementation of set
         @Override
         public T get(Map<ExecutionAttribute<?>, Object> attributes) {
-            return previousValue.updateAndGet(t -> readMapping.apply(t, (U) attributes.get(realAttributeSupplier.get())));
+            return readMapping.apply(
+                (T) attributes.get(backingAttributeSupplier.get()),
+                (U) attributes.get(attributeSupplier.get())
+            );
         }
 
         @SuppressWarnings("unchecked") // Safe because of the implementation of set
         @Override
         public void set(Map<ExecutionAttribute<?>, Object> attributes, T value) {
-            previousValue.set(value);
-            attributes.compute(realAttributeSupplier.get(), (k, real) -> writeMapping.apply((U) real, value));
+            attributes.compute(backingAttributeSupplier.get(), (k, attr) -> value);
```

Comment:
Nit: Is this just `attributes.put(backingAttributeSupplier.get(), value);`

Reply:
Yeah.